File: src/lib/scrapers/bol.ts
import * as cheerio from "cheerio";
import type { PriceResult } from "../types";

const BASE = "https://www.bol.com";

/**
 * Best-effort scraper: bol.com heeft sterke bot-detectie (consent-wall /
 * client-side rendering). Vanaf datacenter-IP's levert dit vaak 0 resultaten
 * op; de zoekroute valt dan terug op demo-data totdat de Marketing Catalog
 * API beschikbaar is (KvK vereist).
 */
export async function searchBol(query: string): Promise<PriceResult[]> {
  const url = `${BASE}/nl/nl/s/?searchtext=${encodeURIComponent(query)}`;

  const res = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
      "Accept-Language": "nl-NL,nl;q=0.9",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    },
    signal: AbortSignal.timeout(8000),
    next: { revalidate: 300 },
  });

  if (!res.ok) throw new Error(`Bol.com HTTP ${res.status}`);

  const html = await res.text();
  const $ = cheerio.load(html);
  const results: PriceResult[] = [];

  // Elke product-card is een div[role="button"] met daarin een /nl/nl/p/-link
  $('div[role="button"]').each((_, el) => {
    const titleLink = $(el)
      .find('a[href^="/nl/nl/p/"]')
      .filter((_, a) => $(a).text().trim().length > 10)
      .first();
    const name = titleLink.text().trim();
    const href = titleLink.attr("href") ?? "";
    const link = href.startsWith("http") ? href : `${BASE}${href}`;

    // Screen-reader span: "De prijs van dit product is '2261' euro en '00' cent"
    const srPrice = $(el)
      .find("span")
      .filter((_, s) => $(s).text().startsWith("De prijs van dit product"))
      .first()
      .text();
    const m = srPrice.match(/'(\d+)' euro en '(\d+)' cent/);
    const price = m ? Number(m[1]) + Number(m[2]) / 100 : NaN;

    // Productafbeelding staat in de parent-wrapper naast de content-kolom
    const img = $(el).parent().find('img[src*="media.s-bol"]').first().attr("src");

    const cardText = $(el).text();
    const inStock = !/uitverkocht|niet leverbaar/i.test(cardText);

    if (name && !isNaN(price) && price > 0) {
      results.push({ retailer: "bol", name, priceEur: price, url: link, imageUrl: img, inStock });
    }
  });

  return results.slice(0, 10);
}

File: scrapers/corebuild_scrapers/retailers/bol.py
"""Bol.com: sterke bot-detectie — werkt vanaf residentiële IP's.

Prijs staat in een screen-reader-span:
"De prijs van dit product is '2261' euro en '00' cent".
"""

import re
from urllib.parse import quote

from bs4 import BeautifulSoup

from .common import MAX_RESULTS, get

BASE = "https://www.bol.com"

PRICE_RE = re.compile(r"'(\d+)' euro en '(\d+)' cent")


def search(query: str) -> list[dict]:
    res = get(f"{BASE}/nl/nl/s/?searchtext={quote(query)}")
    soup = BeautifulSoup(res.text, "html.parser")
    results = []

    for card in soup.select('div[role="button"]'):
        # Titel-link: de product-link met leesbare tekst
        title_link = None
        for a in card.select('a[href^="/nl/nl/p/"]'):
            if len(a.get_text(strip=True)) > 10:
                title_link = a
                break
        if not title_link:
            continue

        sr = card.find("span", string=PRICE_RE)
        m = PRICE_RE.search(sr.get_text()) if sr else None
        if not m:
            # fallback: zoek in alle span-teksten
            for span in card.find_all("span"):
                m = PRICE_RE.search(span.get_text())
                if m:
                    break
        if not m:
            continue
        price = int(m.group(1)) + int(m.group(2)) / 100

        href = title_link.get("href", "")
        parent = card.parent
        img = parent.select_one('img[src*="media.s-bol"]') if parent else None
        in_stock = not re.search(r"uitverkocht|niet leverbaar", card.get_text(" "), re.I)

        results.append(
            {
                "retailer": "bol",
                "name": title_link.get_text(strip=True),
                "price_eur": price,
                "url": href if href.startswith("http") else f"{BASE}{href}",
                "image_url": img.get("src") if img else None,
                "in_stock": in_stock,
            }
        )
        if len(results) >= MAX_RESULTS:
            break

    return results
